### ISSUE TITLE
Testing: REST API compatibility checker

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#2649]
Within major versions we should retain compatibility for REST API clients. This means not removing fields and keeping values on the same format. (I think adding fields is allowed in JSON?)

To enforce this, it would be nice if the unit test suite had a tool that checked through requests and compared to known return formats. This would probably be best accomplished by slightly refactoring the GUI code to take a couple of interfaces that can be mocked out.

**Comment by calmh** [comment]
Input from @Zillode, @Nutomic, @canton7 as the most regular API consumers on what needs to be checked for. Can we add fields? Somehow declare valid values on string fields that are actually enums? Things like that?

**Comment by canton7** [comment]
Adding fields is fine. Quasi-enums can be documented, and if you document them as potentially gaining new values in the future I can ensure that that won't break things.

**Comment by calmh** [comment]
It would be neat if we could document the expected formats as comments on the actual methods in the GUI code, and from that autogenerate tests and the API documentation. Maybe, maybe not.

**Comment by canton7** [comment]
There are various json schema specs as well. Maybe document responses using a schema, and test them against the schema? You can probably also gen docs from the schema.

**Comment by calmh** [comment]
Indeed! http://json-schema.org/examples.html as the first google hit.

**Comment by canton7** [comment]
Looks like there's a [go validator](https://github.com/xeipuuv/gojsonschema) as well.

**Comment by AudriusButkevicius** [comment]
Must have clicked accidentially

**Comment by lkwg82** [comment]
IMHO best way to is to introduce api versioning, else you add complexity, because you can not refactor your api.

This introduces the overhead of building an internal abstraction layer .... so much written on the web.
Maybe we can look into dockers way of implementing this: https://docs.docker.com/engine/reference/api/docker_remote_api/

**Comment by canton7** [comment]
I think there are separate questions here. Ensuring you don't break one version of an API, and introducing different versions of an API, are different but related tasks.

**Comment by calmh** [comment]
I have no desire to maintain multiple api versions; we'll break the api when we need to, but we should do so only intentionally and at major release points.

**Comment by Nutomic** [comment]
We should also assure that types stay the same, eg an int field does not become a string. Adding fields is fine for the Android client.

**Comment by lkwg82** [comment]
Any objections against closing this issue?

**Comment by canton7** [comment]
It still seems like a valid suggestion for an enhancement. Just because something is low down on the priority list, doesn't mean it isn't *on* the list, somewhere.
